### PR TITLE
[Labs] new TagInput component!

### PR DESCRIPTION
**Pull request by giladgray**
#### Changes proposed in this pull request:

`TagInput` renders `Tag`s inside an input, followed by an actual text input. The container is merely styled to look like a Blueprint input; the actual editable element appears after the last tag. Clicking anywhere on the container will focus the text input for seamless interaction.

#### Reviewers should focus on:
- API design. `onAdd` and `onRemove` seem sufficient in my testing--i was able to implement a basic multi-select with little trouble.

#### Screenshot

image

### CODE REVIEW

**Comment by cmslewis** [comment]
Stress testing (not much we should do to mitigate this maybe, just FYI):

File: packages/labs/src/tagInput/tagInput.tsx [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+                    ref={this.refHandlers.input}
+                    className={classNames("pt-input-invisible", inputProps.className)}
+                />
+            </div>
+        );
+    }
+
+    private handleContainerClick = () => {
+        if (this.inputElement != null) {
+            this.inputElement.focus();
```
Comment (cmslewis) [inline packages/labs/src/tagInput/tagInput.tsx]: If I'm blurred and then click a tag's remove button, is it still expected that the input focus?

File: packages/labs/src/tagInput/tagInput.tsx [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+    private handleInputChange = (event: React.KeyboardEvent<HTMLInputElement>) => {
+        this.setState({ inputValue: event.currentTarget.value });
+        Utils.safeInvoke(this.props.inputProps.onChange, event);
+    }
+
+    private handleInputKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
+        const { value } = event.currentTarget;
+        if (event.which === Keys.ENTER && value.length > 0) {
+            Utils.safeInvoke(this.props.onAdd, value);
```
Comment (cmslewis) [inline packages/labs/src/tagInput/tagInput.tsx]: Elsewhere, we `setState` before invoking the callback. Is the switch here intentional?

File: packages/labs/src/tagInput/tagInput.tsx [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+            this.setState({ inputValue: "" });
+        } else if (event.which === Keys.BACKSPACE && value.length === 0) {
+            const lastIndex = this.props.values.length - 1;
+            Utils.safeInvoke(this.props.onRemove, this.props.values[lastIndex], lastIndex);
+        }
+        Utils.safeInvoke(this.props.inputProps.onKeyDown, event);
+    }
+
+    private handleRemoveTag = (event: React.MouseEvent<HTMLSpanElement>) => {
+        // using data attribute to simplify callback logic -- one handler for all children
```
Comment (cmslewis) [inline packages/labs/src/tagInput/tagInput.tsx]: :1st_place_medal:

**Comment by llorca** [comment]
Note: aware that `.pt-tag.pt-active` isn't great visually, I'll push better styles for it later

File: packages/labs/src/tagInput/tagInput.tsx [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    };
+
+    public state: ITagInputState = { inputValue: "", isInputFocused: false };
+
+    private inputElement: HTMLInputElement;
+    private refHandlers = {
+        input: (ref: HTMLInputElement) => {
+            this.inputElement = ref;
+            const refHandler = this.props.inputProps.ref;
+            if (Utils.isFunction(refHandler)) {
```
Comment (giladgray) [inline packages/labs/src/tagInput/tagInput.tsx]: seems there's not much to be done to improve `safeInvoke` usage. if i make it accept non-functions then it loses all type safety :(. adding code comment instead.

**Pull request merged**